File: 100_same_tree/solution.py
from typing import Optional


class Node:
    def __init__(self, val=None):
        self.val = val
        self.left = None
        self.right = None


root1 = Node(1)
root1.left = Node(2)
root1.right = Node(3)

root2 = Node(1)
root2.left = Node(2)
root2.right = Node(3)


class Solution:
    def isSameTree(self, p: Optional[Node], q: Optional[Node]) -> bool:
        if not p and not q:
            return True
        if p and not q:
            return False
        if q and not p:
            return False
        return p.val == q.val and self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)


print(Solution().isSameTree(root1, root2))

File: 100_same_tree/solution.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSameTree(p *TreeNode, q *TreeNode) bool {
	if p == nil && q == nil {
		return true
	}
	if (p == nil && q != nil) || (p != nil && q == nil) {
		return false
	}
	return p.Val == q.Val && isSameTree(p.Left, q.Left) && isSameTree(p.Right, q.Right)
}

func main() {
	tree1 := TreeNode{
		Val:   1,
		Left:  &TreeNode{Val: 2},
		Right: &TreeNode{Val: 3, Right: &TreeNode{Val: 4}},
	}
	tree2 := TreeNode{
		Val:   1,
		Left:  &TreeNode{Val: 2},
		Right: &TreeNode{Val: 3, Right: &TreeNode{Val: 5}},
	}
	fmt.Println(isSameTree(&tree1, &tree2))
}
